pkg/maven/settings.go

```diff
@@ -0,0 +1,150 @@
+package maven
+
+import (
+	"errors"
+	"fmt"
+	piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/piperutils"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+var getenv = os.Getenv
+
+// SettingsFileType ...
+type SettingsFileType int
+
+const (
+	// GlobalSettingsFile ...
+	GlobalSettingsFile SettingsFileType = iota
+	// ProjectSettingsFile ...
+	ProjectSettingsFile SettingsFileType = iota
+)
+
+// GetSettingsFile ...
+func GetSettingsFile(settingsFileType SettingsFileType, src string, fileUtils piperutils.FileUtils, httpClient piperhttp.Downloader) error {
+
+	var dest string
+	var err error
+
+	switch settingsFileType {
+	case GlobalSettingsFile:
+		dest, err = getGlobalSettingsFileDest()
+	case ProjectSettingsFile:
+		dest, err = getProjectSettingsFileDest()
+	default:
+		return errors.New("Invalid SettingsFileType")
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if len(src) > 0 {
+
+		log.Entry().Debugf("Copying file \"%s\" to \"%s\"", src, dest)
+
+		if strings.HasPrefix(src, "http:") || strings.HasPrefix(src, "https:") {
+
+			if err := httpClient.DownloadFile(src, dest, nil, nil); err != nil {
+				return err
+			}
+		} else {
+
+			// for sake os symetry it would be better to use a file protocol prefix here (file:)
+
+			parent := filepath.Dir(dest)
+
+			exists, err := fileUtils.FileExists(parent)
+
+			if err != nil {
+				return err
+			}
+
+			if !exists {
+				if err = fileUtils.MkdirAll(parent, 0775); err != nil {
+					return err
+				}
+			}
+
+			exists, err = fileUtils.FileExists(src)
+
+			if err != nil {
+				return err
+			}
+
+			if !exists {
+				return fmt.Errorf("File \"%s\" not found", src)
+			}
+
+			fmt.Printf("Copying file '%s' to '%s'", src, dest)
+			if _, err := fileUtils.Copy(src, dest); err != nil {
```

Comment:
... I'm also new to go and collecting experience. I don't get the point here. Above I need more lines of code since I first check the error and afterwards I check if the file exists. Here I copy the file - after we checked that the file exists.

Would you like to omit the file exists check? In general we can do so. Depends to some extend to programming habits. IMO it is a good practice to check first if a file exists, and if it exists we copy it. But I'm not strongly against omitting the prior file exists check. E.g. it might happen that a file is not readable for the current user even if it exists (not seen that in  practice so far, but possible).

Reply:
Just a suggestion: At least for the copying code, you could also just try to copy without checking if the source exist and it will just fail much in the same way when the file does not exist. And for fileUtils.MkdirAll(), the question is what it considers as error. I guess there are use-cases where it would be an error if the folder exists (since you wanted a clean start), but in many other use-cases you just want to make sure the path exists. This could be reflected in parameters to MkdirAll() and then it would return no error if the path already exists. This would greatly simplify the code here.

Reply:
Ok, i will shorten the code and not check if the source file exists. Here for this use case it is better to check for intermediate directories and create these directories. Example: I can imagine the is maven installed, but for the current user there is no `~/.m2`. In that case ~~there is nothing wrong with~~ it is the expected behaviour to create that folder. It is not a folder structure in that case, but `MkdirsAll` is the more general approach compared to only create on level.